File: src/utils/math/SingleBoundedFunction.java
package utils.math;

import com.zypex.piflow.profile.Linear;

public class SingleBoundedFunction<T> implements BoundedFunction<T>{

    private final double upper;
    private final double lower;
    private final Function<T> base;

    public SingleBoundedFunction(Function<T> baseFunction, double lower, double upper) {
        this.base = baseFunction;
        this.lower = lower;

        this.upper = upper;
    }

//    @Override
//    public <R extends BoundedFunction<T>> R offset(double offset) {
//        return null;
//    }

    @Override
    public SingleBoundedFunction<T> offset(double offset) {
        return new SingleBoundedFunction<>(t -> base.get(t - offset), lower + offset, upper + offset);
    }

    @Override
    public double upperBound() {
        return upper;
    }

    @Override
    public double lowerBound() {
        return lower;
    }

    @Override
    public T get(double input) {
        if(input >= lower && input <= upper) return base.get(input);
        throw new InputOutOfDomainException("Input " + input + " out of domain [" + lower + ", " + upper + "]");
    }


}

File: src/utils/math/SingleBoundedFunction.kt
package utils.math


class SingleBoundedFunction<T>(private val base: (Double) -> T,
                               private val lowerBound: Double,
                               private val upperBound: Double,
) : BoundedFunction<T>{

    override fun offset(offset: Double): SingleBoundedFunction<T> = SingleBoundedFunction({ t -> base(t - offset) }, lowerBound + offset, upperBound + offset)
//    override val upper: T
//        get() = base(upperBound)
//
//    override val lower: T
//        get() = base(lowerBound)

    override fun upperBound(): Double = upperBound
    override fun lowerBound(): Double = lowerBound

    override fun invoke(input: Double): T {
        if (input in lowerBound..upperBound) return base(input)
        throw InputOutOfDomainException("Input $input out of domain [$lowerBound, $upperBound]")
    }
}